Update to permit multiple root dirs

CBroz1 opened the pull request:
Changed:
- Updated mosts tests with docstrings.
- PEP8
   - line length to ≤ 88
   - removed unused dependencies
   - full words in variable/function names
- Include `element-interface.utils` path functions
- Rebased with main commit history
- Added `verbose` boolean to integration tests for simpler output
- Changed docker.test command for more complete coverage test output

Pending:
- Successful docker integration tests

kabilar commented on workflow_array_ephys/paths.py:
```diff
@@ -4,12 +4,9 @@
-
-    return root_data_dirs
-
+    return root_data_dirs if root_data_dirs else None
```
Would the `get` method in the above statement set `root_data_dirs` to `None` if the key doesn't exist?  In which case, this if-else statement is not necessary.

CBroz1 replied:
That's a good idea. I've added this to the batch of changes.

kabilar commented on tests/test_populate.py:
```diff
@@ -32,6 +34,7 @@ def test_LFP_populate_npx3B_OpenEphys(testdata_paths, pipeline, ephys_recordings
 
 
 def test_LFP_populate_npx3A_SpikeGLX(testdata_paths, pipeline, ephys_recordings):
+    """Populate ephys.LFP with SpikeGLX items, recording npx3A"""
```
```suggestion
    """Populate ephys.LFP with SpikeGLX items, recording Neuropixels Phase 3A probes"""
```

kabilar commented on tests/test_populate.py:
```diff
@@ -52,6 +55,8 @@ def test_LFP_populate_npx3A_SpikeGLX(testdata_paths, pipeline, ephys_recordings)
 
 
 def test_LFP_populate_npx3B_SpikeGLX(testdata_paths, pipeline, ephys_recordings):
+    """Populate ephys.LFP with SpikeGLX items, recording npx3B"""
```
```suggestion
    """Populate ephys.LFP with SpikeGLX items, recording Neuropixels Phase 3B (Neuropixels 1.0)"""
```

kabilar commented on tests/test_populate.py:
```diff
@@ -99,8 +105,8 @@ def test_curated_clustering_populate(curations, pipeline, testdata_paths):
 
 
 def test_waveform_populate_npx3B_OpenEphys(curations, pipeline, testdata_paths):
+    """Populate ephys.WaveformSet with OpenEphys npx3B"""
```
```suggestion
    """Populate ephys.WaveformSet with OpenEphys Neuropixels Phase 3B (Neuropixels 1.0) probe"""
```

kabilar commented on tests/test_populate.py:
```diff
@@ -13,6 +13,8 @@ def test_ephys_recording_populate(pipeline, ephys_recordings):
 
 
 def test_LFP_populate_npx3B_OpenEphys(testdata_paths, pipeline, ephys_recordings):
+    """Populate ephys.LFP with OpenEphys items"""
```
```suggestion
    """Populate ephys.LFP with OpenEphys items, recording Neuropixels Phase 3B (Neuropixels 1.0) probe"""
```

kabilar commented on tests/test_populate.py:
```diff
@@ -113,6 +119,8 @@ def test_waveform_populate_npx3B_OpenEphys(curations, pipeline, testdata_paths):
 
 
 def test_waveform_populate_npx3B_SpikeGLX(curations, pipeline, testdata_paths):
+    """Populate ephys.WaveformSet with SpikeGLX npx3B"""
```
```suggestion
    """Populate ephys.WaveformSet with SpikeGLX Neuropixels Phase 3B (Neuropixels 1.0) probe"""
```

kabilar commented on workflow_array_ephys/ingest.py:
```diff
@@ -9,38 +9,44 @@
     with open(subject_csv_path, newline= '') as f:
         input_subjects = list(csv.DictReader(f, delimiter=','))
-
+    # Broz 102821 - this gives full # even if skipped, not # populated
```
```suggestion
```

Should we delete this note?

kabilar commented on workflow_array_ephys/ingest.py:
```diff
@@ -9,38 +9,44 @@
 
 def ingest_subjects(subject_csv_path='./user_data/subjects.csv'):
+    """
+    Ingest subjects listed in the subject column of of ./user_data/subjects.csv
```
```suggestion
    Ingest subjects listed in the subject column of ./user_data/subjects.csv
```

kabilar commented on workflow_array_ephys/ingest.py:
```diff
@@ -71,18 +77,22 @@ def ingest_sessions(session_csv_path='./user_data/sessions.csv'):
-    ephys.ProbeInsertion.insert(probe_insertion_list)
+    ephys.ProbeInsertion.insert(probe_insertion_list, skip_duplicates=True)
 
     print('\n---- Successfully completed workflow_array_ephys/ingest.py ----')
```
```suggestion
    print('\n---- Successfully completed ingest_subjects ----')
```

kabilar commented on workflow_array_ephys/ingest.py:
```diff
@@ -71,18 +77,22 @@ def ingest_sessions(session_csv_path='./user_data/sessions.csv'):
     print(f'\n---- Insert {len(session_list)} entry(s) into session.Session ----')
-    session.Session.insert(session_list)
-    session.SessionDirectory.insert(session_dir_list)
+    # Broz 102821 - prev, skip_dupes was true for ingest_subj, but not ingest_sess
+    #               I thought they should mirror each other, chose both True
```
```suggestion
```
These comments are great to understand your thought process.  Please remove from the code and add it to the `Conversation` tab in the pull request.  Thanks.

kabilar replied:
For `session.Session` and `session.SessionDirectory`, I don't believe `skip_duplicates=True` is required because of the following line:
```
if session_key not in session.Session():
```

kabilar commented on workflow_array_ephys/ingest.py:
```diff
@@ -9,38 +9,44 @@
         input_subjects = list(csv.DictReader(f, delimiter=','))
-
+    # Broz 102821 - this gives full # even if skipped, not # populated
     print(f'\n---- Insert {len(input_subjects)} entry(s) into subject.Subject ----')
```
```suggestion
    print(f'\n---- Insert {len(set(input_subjects))} entry(s) into subject.Subject ----')
```

kabilar commented on workflow_array_ephys/ingest.py:
```diff
@@ -9,38 +9,44 @@
     for sess in input_sessions:
-        session_dir = element_data_loader.utils.find_full_path(
-                                                    get_ephys_root_data_dir(), 
+        sess_dir = element_data_loader.utils.find_full_path(
```
```suggestion
        session_dir = element_data_loader.utils.find_full_path(
```

I would suggest keeping the variable name `session_dir`.

CBroz1 replied:
Sure - I'll make sure this is consistent across the workflow

kabilar commented on workflow_array_ephys/ingest.py:
```diff
@@ -9,38 +9,44 @@
 
     # Folder structure: root / subject / session / probe / .ap.meta
-    session_list, session_dir_list, probe_list, probe_insertion_list = [], [], [], []
+    session_list, sess_dir_list, probe_list, probe_insertion_list = [], [], [], []
```
```suggestion
    session_list, session_dir_list, probe_list, probe_insertion_list = [], [], [], []
```

kabilar commented on workflow_array_ephys/ingest.py:
```diff
@@ -71,18 +77,22 @@ def ingest_sessions(session_csv_path='./user_data/sessions.csv'):
 
     print(f'\n---- Insert {len(probe_list)} entry(s) into probe.Probe ----')
-    probe.Probe.insert(probe_list)
+    probe.Probe.insert(probe_list, skip_duplicates=True)
```
For `probe.Probe`, I don't believe that `skip_duplicates=True` is required because of the following line:
```
if probe_key['probe'] not in [p['probe'] for p in probe_list] and probe_key not in probe.Probe():
    probe_list.append(probe_key)
```

dimitri-yatsenko replied:
Yes, generally, do not use `skip_duplicates`. Catching duplication errors may be important for some integrity checks.

CBroz1 replied:
Good note! I now see why this isn't needed, thanks. And yes, certainly best practice to leave as default `skip_duplicates=False`

kabilar commented on tests/test_ingest.py:
```diff
@@ -9,6 +9,7 @@
 
 
 def test_ingest_subjects(pipeline, ingest_subjects):
+    """Check length of subject.Subject"""
```
```suggestion
    """ Check number of subjects inserted into the `subject.Subject` table """
```

kabilar commented on tests/__init__.py:
```diff
@@ -240,6 +252,7 @@ def clustering_tasks(pipeline, kilosort_paramset, ephys_recordings):
 
 @pytest.fixture
 def clustering(clustering_tasks, pipeline):
+    """Populate ehys.Clustering"""
```
```suggestion
    """Populate ephys.Clustering"""
```

kabilar commented on tests/__init__.py:
```diff
@@ -153,6 +163,7 @@ def ingest_sessions(ingest_subjects, sessions_csv):
 
 @pytest.fixture
 def testdata_paths():
+    """ Paths for testdata 'subjX/sessY/probeZ/etc'"""
```
```suggestion
    """ Paths for test data 'subjectX/sessionY/probeZ/etc'"""
```

kabilar commented on tests/test_ingest.py:
```diff
@@ -69,7 +76,7 @@ def test_find_root_directory(pipeline, sessions_csv):
     root_dir = find_root_directory(ephys_root_data_dir, session_full_path)
 
     assert root_dir.as_posix() == get_ephys_root_data_dir()
```
```suggestion
    session_full_path = pathlib.Path('/main/workflow-array-ephys/tests/user_data') / sess.session_dir
    root_dir = find_root_directory(ephys_root_data_dir, session_full_path)

    assert root_dir.as_posix() == '/main/workflow-array-ephys/tests/user_data'
```

kabilar replied:
Add remove the previous line
```
session_full_path = pathlib.Path(get_ephys_root_data_dir()) / sess.session_dir
```

kabilar replied:
Modifying this test should still allow for testing the `find_root_directory` function.  Do you think this works correctly?

CBroz1 replied:
That makes sense - I think it should work. We would still test that the data was downloaded with the relative directories as intended. I'll run the tests to be sure.

kabilar commented on tests/test_ingest.py:
```diff
@@ -29,6 +30,12 @@ def test_ingest_sessions(pipeline, sessions_csv, ingest_sessions):
     assert (session.SessionDirectory
             & {'subject': sess.name}).fetch1('session_dir') == sess.session_dir
 
+''' Delete these?
+CB: I think these tests are depreciated with the update to permit multiple root directories
+    Previously, they tested against known bad roots to make sure root/session matched
+    Now, we have to run find_full_path every on mult roots regardless.
+    To update would result in tautology:
+    > assert find_full_path == find_full_path
```
```suggestion
```

In the `test_find_valid_full_path` function, the following should still allow for testing the `find_full_path` function by changing the line from:
```
session_full_path = pathlib.Path(get_ephys_root_data_dir()) / sess.session_dir
```

to:
```
session_full_path = pathlib.Path('/main/workflow-array-ephys/tests/user_data') / sess.session_dir
```

kabilar commented on tests/__init__.py:
```diff
@@ -166,6 +177,7 @@ def testdata_paths():
 
 @pytest.fixture
 def kilosort_paramset(pipeline):
+    """Insert kilosort params into ephys.ClusteringParamset"""
```
```suggestion
    """Insert kilosort parameters into ephys.ClusteringParamset"""
```

kabilar commented on tests/__init__.py:
```diff
@@ -1,14 +1,17 @@
-# run tests: pytest -sv --cov-report term-missing --cov=workflow-array-ephys -p no:warnings
+# dependencies: pip install pytest pytest-cov
```
```suggestion
```

Duplicate from `requirements_test.txt`.

kabilar commented on tests/__init__.py:
```diff
@@ -1,14 +1,17 @@
 
 import workflow_array_ephys
 from workflow_array_ephys.paths import get_ephys_root_data_dir
+import element_data_loader.utils
```
```suggestion
from element_interface.utils import find_root_directory, find_full_path
```
With the recent package rename let's update from `element_data_loader` to `element_interface`.  And to keep consistent with the `elements`, let's import the function from the module.  Please update these scripts to incorporate these changes.

kabilar commented on tests/__init__.py:
```diff
@@ -46,12 +49,18 @@ def dj_config():
-    test_data_dir = pathlib.Path(dj.config['custom']['ephys_root_data_dir'])
-
-    test_data_exists = np.all([(test_data_dir / p).exists() for p in sessions_dirs])
+    """If data not exist, attempt download with DJArchive
+    * If no or partial data present, download to first listed root"""
```
```suggestion
    """If data does not exist or partial data is present, 
    attempt download with DJArchive to the first listed root directory"""
```

kabilar reviewed (changes requested):
Thanks @CBroz1!  I appreciate you updating the workflow `ingest.py` and `pytests` to permit multiple root data directories.

dimitri-yatsenko commented on workflow_array_ephys/paths.py:
```diff
@@ -4,12 +4,9 @@
-
-    return root_data_dirs
-
+    return root_data_dirs if root_data_dirs else None
```
```suggestion
    return root_data_dirs or None
```

CBroz1 replied:
Resolved this function to the following during the rebase process
`return dj.config.get('custom', {}).get('ephys_root_data_dir', None)`

CBroz1 commented:
- Rebased changes made since opening PR
- Troubleshooting docker integration testing:
   - Dockerfile:26 `RUN pip install /main/workflow-array-ephys`
   - Traceback: `OSError: [Errno 28] No space left on device: '/tmp/pip-ephem-wheel-cache-3fjal6tf'`

CBroz1 commented:
See 735063a commit notes above - docker passes with data across multiple roots

CBroz1 commented:
Rebased relative to main public branch. Passes dockerized integration tests. Relies on uploading of example data with split root directory

kabilar commented on README.md:
```diff
@@ -45,7 +45,7 @@ https://github.com/datajoint/element-animal/blob/main/images/subject_diagram.svg
-workflow ([03-process.ipynb](notebooks/03-process.ipynb)) and explore the data 
-([05-explore.ipynb](notebooks/05-explore.ipynb)).
\ No newline at end of file
++ Please refer to the following workflow-specific
+workflow ([03-process.ipynb](notebooks/03-process.ipynb)) and explore the data
+[Jupyter notebooks](/notebooks) for an in-depth explanation of how to run the
+([05-explore.ipynb](notebooks/05-explore.ipynb)).
```
```suggestion
+ Please refer to the following workflow-specific 
 [Jupyter notebooks](/notebooks) for an in-depth explanation of how to run the 
 workflow ([03-process.ipynb](notebooks/03-process.ipynb)) and explore the data 
 ([05-explore.ipynb](notebooks/05-explore.ipynb)).
```

kabilar commented:
Hi @CBroz1, the Docker and Compose files are showing up as new files in this pull request but these files exist in the base repository.  Perhaps the rebase did not take?  Fixing this issue would hopefully fix the conflicts listed above.

CBroz1 commented:
Hi @kabilar - Thanks for your review. Maybe the conflicts were introduced since I rebased last month? Not sure. Fixed now.

I've also updated our `djarchive` bucket to the following
```
<bucket>/workflow-array-ephys-benchmark/0.1.0a4/workflow_ephys_data1/<all except subject4>
<bucket>/workflow-array-ephys-benchmark/0.1.0a4/workflow_ephys_data2/<subject4>

kabilar commented:
Thanks @CBroz1.  I believe the `djarchive-manifest.csv` will need to be within `<bucket>/workflow-array-ephys-benchmark/0.1.0a4/`.  And `djarchive-manifest.csv` will need to be updated for the new subdirectories (`workflow_ephys_data1`, `workflow_ephys_data2`).  And lets rename the revision `0.1.0a4` to 'v2'.

kabilar commented on workflow_array_ephys/ingest.py:
```diff
@@ -1,90 +1,127 @@
+            probe_insertion_list.extend([{**session_key, **insertion
+                                          } for insertion in insertions])
+
+    if verbose:
+        previous_length = len(session.Session.fetch())
     session.Session.insert(session_list)
     session.SessionDirectory.insert(session_dir_list)
+    if verbose:
+        insert_length = len(session.Session.fetch()) - previous_length
```
Perhaps we don't need this logic because of the above statement: `if session_key not in session.Session()`?

kabilar replied:
Probably the same is the case for the tables below (`probe.Probe`, `ephys.ProbeInsertion`).

CBroz1 replied:
You're right - I over-applied this logic. Good catch, thanks

kabilar commented on workflow_array_ephys/ingest.py:
```diff
@@ -1,90 +1,118 @@
-    print(f'\n---- Insert {len(probe_list)} entry(s) into probe.Probe ----')
     probe.Probe.insert(probe_list)
+    if verbose:
+        print(f'\n---- Insert {probe_list} entry(s) into probe.Probe ----')
```
```suggestion
        print(f'\n---- Insert {len(probe_list)} entry(s) into probe.Probe ----')
```

CBroz1 commented:
> * I believe the `djarchive-manifest.csv` file will need to be within `<bucket>/workflow-array-ephys-benchmark/0.1.0a4/`.
> * The `djarchive-manifest.csv` file will need to be updated for the new subdirectories (`workflow_ephys_data1`, `workflow_ephys_data2`).
> * Let's rename the revision `0.1.0a4` to `v2`.

I've added the manifest file, but will hold off on renaming pending [dev request #20](https://github.com/vathes/science-team/issues/20).

I've also changed `tests/__init__.py` to reflect the hardcoded v2 versioning for this test dataset.

kabilar commented on docker-compose-test.yaml:
```diff
@@ -22,20 +25,22 @@ services:
     volumes:
       - ${TEST_DATA_DIR}:/main/test_data
       - ./apt_requirements.txt:/tmp/apt_requirements.txt
+      - ../element-array-ephys:/main/element-array-ephys
```
```suggestion
      - ../element-lab:/main/element-lab
      - ../element-animal:/main/element-animal
      - ../element-session:/main/element-session
      - ../element-array-ephys:/main/element-array-ephys
```

kabilar commented on Dockerfile.test:
```diff
@@ -8,20 +8,31 @@ USER anaconda
+# RUN pip install git+https://github.com/<user>/element-animal.git
+# RUN pip install git+https://github.com/<user>/element-session.git
 # RUN pip install git+https://github.com/<user>/element-array-ephys.git
-# RUN git clone https://github.com/<user>/workflow-array-ephys.git /main/workflow-array-ephys
```
I think we would need this clone for option 2.

kabilar commented on tests/__init__.py:
```diff
@@ -26,68 +31,109 @@
-    test_data_exists = np.all([(test_data_dir / p).exists() for p in sessions_dirs])
+    """If data does not exist or partial data is present,
+    attempt download with DJArchive to the first listed root directory"""
+    test_data_dirs = []
+    test_data_exists = True
+    for p in sessions_dirs:
+        try:
+            test_data_dirs.append(find_full_path(get_ephys_root_data_dir(), p))
+        except FileNotFoundError:
+            test_data_exists = False   # If data not found
```
```suggestion
    test_data_exists = True
    for p in sessions_dirs:
        try:
           find_full_path(get_ephys_root_data_dir(), p)
        except FileNotFoundError:
            test_data_exists = False   # If data not found
```

kabilar replied:
Since `test_data_dirs` doesn't seem to be used in this function, perhaps we can just call `find_full_path`?

kabilar commented on tests/__init__.py:
```diff
@@ -26,68 +31,109 @@
         import djarchive_client
         client = djarchive_client.client()
-        workflow_version = workflow_array_ephys.version.__version__
+        workflow_version = 'v2'
```
```suggestion
```

kabilar commented on tests/__init__.py:
```diff
@@ -26,68 +31,109 @@
-           'lab': pipeline.lab,
-           'ephys': pipeline.ephys,
-           'probe': pipeline.probe,
-           'session': pipeline.session,
-           'get_ephys_root_data_dir': pipeline.get_ephys_root_data_dir}
+        yield {'subject': pipeline.subject,
+               'lab': pipeline.lab,
+               'ephys': pipeline.ephys,
+               'probe': pipeline.probe,
+               'session': pipeline.session,
+               'get_ephys_root_data_dir': pipeline.get_ephys_root_data_dir}
```
What output does this block create?  Wondering if we can refactor this block out of the if/else statement.

CBroz1 replied:
The duplication currently under the `pipeline` fixture is motivated the verbosity option I added. If not verbose, the pipeline is constructed under an instance of the `QuietStdOut` class. I was motivated to add a verbosity flag after running the pytests several times and having difficulty visually parsing the errors. 

There are probably other ways to quiet standard output, but this seemed to be the cleanest of the options I found. I would welcome a proposed alternative

CBroz1 replied:
@kabilar - yes, you were right. functionality was the same if the verbose conditional was only applied to the teardown

kabilar commented on tests/test_ingest.py:
```diff
@@ -34,50 +39,68 @@ def test_find_valid_full_path(pipeline, sessions_csv):
     from element_interface.utils import find_full_path
 
     get_ephys_root_data_dir = pipeline['get_ephys_root_data_dir']
+    if not isinstance(get_ephys_root_data_dir(), list):    # ensure is list
+        ephys_root_data_dir = [get_ephys_root_data_dir()]  # for appending below
+    else:
+        ephys_root_data_dir = get_ephys_root_data_dir()
```
```suggestion
    ephys_root_data_dir = [get_ephys_root_data_dir()] if not isinstance(get_ephys_root_data_dir(), list) else get_ephys_root_data_dir()
```

kabilar commented on tests/test_ingest.py:
```diff
@@ -34,50 +39,68 @@ def test_find_valid_full_path(pipeline, sessions_csv):
     sessions, _ = sessions_csv
     sess = sessions.iloc[0]
-    session_full_path = pathlib.Path(get_ephys_root_data_dir()) / sess.session_dir
+    docker_full_path = find_full_path(['/main/test_data/workflow_ephys_data1/',
+                                       '/main/test_data/workflow_ephys_data2/'],
+                                      sess.session_dir)
```
```suggestion
    docker_full_path = pathlib.Path('/main/test_data/workflow_ephys_data1/') / sess.session_dir
```

kabilar commented on tests/test_ingest.py:
```diff
@@ -34,50 +39,68 @@ def test_find_valid_full_path(pipeline, sessions_csv):
     from element_interface.utils import find_root_directory
 
     get_ephys_root_data_dir = pipeline['get_ephys_root_data_dir']
+    if not isinstance(get_ephys_root_data_dir(), list):    # ensure is list
+        ephys_root_data_dir = [get_ephys_root_data_dir()]  # for appending below
+    else:
+        ephys_root_data_dir = get_ephys_root_data_dir()
```
```suggestion
    ephys_root_data_dir = [get_ephys_root_data_dir()] if not isinstance(get_ephys_root_data_dir(), list) else get_ephys_root_data_dir()
```

kabilar reviewed (changes requested):
Thanks @CBroz1.  Great work.  Few suggestions above.  We are close to merging!

kabilar commented on tests/test_ingest.py:
```diff
@@ -34,50 +39,65 @@ def test_find_valid_full_path(pipeline, sessions_csv):
 
 def test_find_root_directory(pipeline, sessions_csv):
+    """
+    Test that ephys_root_data_dir loaded as docker directory
+        /main/test_data/workflow_ephys_data1/
```
```suggestion
    Test that `find_root_directory` works correctly.
```

kabilar commented on tests/test_ingest.py:
```diff
@@ -34,50 +39,65 @@ def test_find_valid_full_path(pipeline, sessions_csv):
 
-    assert root_dir.as_posix() == get_ephys_root_data_dir()
+    assert root_dir.as_posix() == '/main/test_data/workflow_ephys_data1',\
+        'Root path does not match docker: /main/test_data/workflow_ephys_data1'
```
```suggestion
        'Root path does not match: /main/test_data/workflow_ephys_data1'
```

kabilar commented on tests/__init__.py:
```diff
@@ -26,68 +31,94 @@
-    if _tear_down:
+    if verbose and _tear_down:
         pipeline.subject.Subject.delete()
+    if _tear_down:
```
```suggestion
    elif not verbose and _tear_down:
```

kabilar reviewed (approved):
Thanks @CBroz1!
